library/ecp.c

```diff
@@ -1297,6 +1434,180 @@ static int mbedtls_ecp_sw_derive_y(const mbedtls_ecp_group *grp,
+        return MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+    }
+
+    if (mbedtls_mpi_cmp_int(&pt->Z, 1) != 0) {
+        return MBEDTLS_ERR_MPI_BAD_INPUT_DATA;
+    }
+
+    MBEDTLS_MPI_CHK(mbedtls_mpi_copy(q, &pt->Y));
+
+    /* From 5.1.2, we shall copy LSB of x to the MSB of y */
+    if (mbedtls_mpi_get_bit(&pt->X, 0)) {
+        MBEDTLS_MPI_CHK(mbedtls_mpi_set_bit(q, grp->pbits, 1));
+    }
+
+cleanup:
+    return ret;
+}
+
+int mbedtls_ecp_point_decode(const mbedtls_ecp_group *grp,
+                             mbedtls_ecp_point *pt,
+                             const mbedtls_mpi *q)
+{
+    int ret = MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    mbedtls_mpi u, v, t;
+    mbedtls_mpi_uint r;
+    size_t plen;
+    int x_0;
+
+    if (mbedtls_ecp_get_type(grp) != MBEDTLS_ECP_TYPE_EDWARDS) {
+        return MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+    }
+
+    mbedtls_mpi_init(&u);
+    mbedtls_mpi_init(&v);
+    mbedtls_mpi_init(&t);
+
+    plen = (mbedtls_mpi_bitlen(&grp->P) + 1 + 7) >> 3;
+    MBEDTLS_MPI_CHK(mbedtls_mpi_copy(&pt->Y, q));
+    /* High bit of last digit is the least significant bit of the
+     * x-coordinate. Save it and clear it. */
+    x_0 = mbedtls_mpi_get_bit(&pt->Y, (plen * 8) - 1);
+    MBEDTLS_MPI_CHK(mbedtls_mpi_set_bit(&pt->Y, (plen * 8) - 1, 0));
+
+    /* If the resulting y-coordinate is >= p, decoding fails. */
+    if (mbedtls_mpi_cmp_mpi(&pt->Y, &grp->P) >= 0) {
+        ret = MBEDTLS_ERR_MPI_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    /* To recover the x-coordinates, the curve equation implies
+       x^2 = (y^2 - 1) / (d y^2 - a) (mod p). The denominator is always
+       non-zero mod p. Let u = y^2 - 1 and v = d y^2 - a. */
+
+    MPI_ECP_MUL(&u, &pt->Y,  &pt->Y);
+    MPI_ECP_MUL(&v, &grp->B, &u);
+    MPI_ECP_SUB(&v, &v,      &grp->A);
+    MPI_ECP_SUB_INT(&u, &u, 1);
+
+    /* We use different algorithm to compute the square root of (u/v)
+     * depending on p (mod 8). */
+    MBEDTLS_MPI_CHK(mbedtls_mpi_mod_int(&r, &grp->P, 8));
+
+    if (r == 3 || r == 7) {
+        /* This corresponds to p = 3 (mod 4) and for instance Ed448. *
+         * The candidate root is x = sqrt(u/v) = u (u v)^((p-3)/4) */
+        MPI_ECP_MUL(&pt->X,  &u, &v);
+        MPI_ECP_SUB_INT(&t, &grp->P, 3);
+        MBEDTLS_MPI_CHK(mbedtls_mpi_shift_r(&t, 2));
+        MBEDTLS_MPI_CHK(mbedtls_mpi_exp_mod(&pt->X, &pt->X, &t, &grp->P, NULL));
+        MPI_ECP_MUL(&pt->X, &pt->X, &u);
+
+        /* If v * x^2 = u (mod p), the recovered x-coordinate is x. Otherwise,
+         * no square root exists, and the decoding fails. */
+        MPI_ECP_MUL(&t,   &pt->X,   &pt->X);
+        MPI_ECP_MUL(&t,   &t,      &v);
+        if (mbedtls_mpi_cmp_mpi(&t, &u) != 0) {
+            ret = MBEDTLS_ERR_MPI_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+    } else if (r == 5) {
+        /* This corresponds for instance to Ed25519. The candidate root is
+         * x = sqrt(u/v) = u (u v)^((p-5)/8) */
```

Comment:
From RFC8032,
x = sqrt(u/v) = (u/v)^((p+3)/8) = u * (v^3) * (u * (v^7))^((p-5)/8)
but here is 
x = sqrt(u/v) = u * (u * v)^((p-5)/8),
Is this a bug?

Reply:
The second formula is an optimization. See https://www.rfc-editor.org/errata/eid5758 for a proposed change of RFC 8032 and https://mailarchive.ietf.org/arch/msg/crypto-panel/iiG3qza8szUfNOIvsUY9umRjQqs/ for a proof.